### ISSUE TITLE
Implement dynamic thread pool

### ISSUE THREAD

**Issue by Al2Klimov** [Icinga/icinga2#5269]
While reproducing #5148 I've seen some `boost::thread(...).detach()` in the code.
On heavy load (e.g. API) this causes lots of threads to be created and destroyed.
This is more expensive than re-using existing threads which have "nothing to do" (like in a thread pool).
Spawning (and destroying) less threads at all could also fix #5148.

But a thread pool "as we know" would limit the efficiency.
So I think a dynamic, "queue-less" thread pool would be a good solution for these problems.

## How it would work (read: how I imagine it right now)

There's one "global" DynamicThreadPool, initialized with 0 threads.
Once a task is being "queued" and there are no free threads at the moment, a new one is being spawned.
Once a thread is done with a task, it doesn't get destroyed, but waits for the next one.

~~To speed up the next-free-thread-lookup there should be a multiset with pointers to the threads sorted by whether the thread behind the pointer is busy.~~

**Comment by Al2Klimov** [comment]
Note:
I could even imagine how some of the algos and data structures would have to look like and I could start coding right now.
But this would make no sense as I don't know whether you would accept a such ("fundamental") feature.

**Comment by dnsmichi** [comment]
Put it on hold and focus on other tasks please. That's fundamental changes which need to be discussed and (re)designed.

**Comment by Al2Klimov** [comment]
"other tasks"? Which "other tasks"?
